Make `FileDialog` filtering case insensitive

DevilboxGames opened the pull request:
When using FileDialog to save files on non-case-sensitive file systems (Windows and often macOS), a file may have an uppercase file extension ( `file.TXT` ) but the filter is set as lowercase ( `*.txt ; Text Files` ). The FileDialog will list the file but selecting it to overwrite will not detect that the file matches the filter and append the lowercase extension after the existing uppercase file extension and would result in the selected file being `file.TXT.txt`

As most other operating systems can be case sensitive, my suggestion is to add a `CaseSensitivity` enum with the values of `True`, `False` and `OS Default` to the FileDialog for toggling between using `f.match(str)` and `f.matchn(str)` in the checks which would allow it to either be explicit based on the developer intent or implicit based on the standard of the operating system.

![filedialog-case-sensitivity](https://github.com/godotengine/godot/assets/7022889/d0553db2-073a-47c4-b0d2-34cf05b82040)

Currently if `OS Default` is used then the code uses a case-sensitive check if the OS is not Windows or macOS. Is there a better approach for this? Are there any other supported OS's are not case-sensitive by default?

Attached is an MRP similar to the one provided in the issue, with the addition of an OptionButton for changing between the CaseSensitivity setting when running the project.
[FileDialogCaseSensitivity.zip](https://github.com/godotengine/godot/files/13561697/FileDialogCaseSensitivity.zip)

Fixes #85788

DevilboxGames committed:
Make File Dialog case insensitive

Change FileDialog and EditorFileDialog to use case insensitive string
comparisons when saving files to avoid duplicate file extensions.

DevilboxGames commented on scene/gui/file_dialog.cpp:
```diff
@@ -355,6 +355,8 @@ void FileDialog::_action_pressed() {
 	if (mode == FILE_MODE_SAVE_FILE) {
 		bool valid = false;
+		String osName = OS::get_singleton()->get_name();
+		bool use_case_sensitive_filter = case_sensitive_filter == CASE_SENSITIVITY_TRUE || case_sensitive_filter == CASE_SENSITIVITY_OS_DEFAULT && osName != "Windows" && osName != "macOS";
```
Aha, thanks for the heads up. I expected there was a more appropriate approach but completely missed that. I've amended that now.

AThousandShips commented on scene/gui/file_dialog.cpp:
```diff
@@ -356,6 +356,8 @@ void FileDialog::_action_pressed() {
 	if (mode == FILE_MODE_SAVE_FILE) {
 		bool valid = false;
 
+		bool use_case_sensitive_filter = case_sensitive_filter == CASE_SENSITIVITY_TRUE || case_sensitive_filter == CASE_SENSITIVITY_OS_DEFAULT && dir_access->is_case_sensitive(f);
```
```suggestion
		bool use_case_sensitive_filter = case_sensitive_filter == CASE_SENSITIVITY_TRUE || (case_sensitive_filter == CASE_SENSITIVITY_OS_DEFAULT && dir_access->is_case_sensitive(f));
```
Or
```suggestion
		bool use_case_sensitive_filter = (case_sensitive_filter == CASE_SENSITIVITY_TRUE || case_sensitive_filter == CASE_SENSITIVITY_OS_DEFAULT) && dir_access->is_case_sensitive(f);
```
Depending on what you indented, it's ambiguous

AThousandShips commented on scene/gui/file_dialog.cpp:
```diff
@@ -364,7 +366,7 @@ void FileDialog::_action_pressed() {
 				for (int j = 0; j < flt.get_slice_count(","); j++) {
 					String str = flt.get_slice(",", j).strip_edges();
-					if (f.match(str)) {
+					if (case_sensitive_filter && f.match(str) || !case_sensitive_filter && f.matchn(str)) {
```
```suggestion
					if ((case_sensitive_filter && f.match(str)) || (!case_sensitive_filter && f.matchn(str))) {
```

AThousandShips commented on scene/gui/file_dialog.cpp:
```diff
@@ -383,7 +385,7 @@ void FileDialog::_action_pressed() {
 				for (int j = 0; j < filterSliceCount; j++) {
 					String str = (flt.get_slice(",", j).strip_edges());
-					if (f.match(str)) {
+					if (use_case_sensitive_filter && f.match(str) || !use_case_sensitive_filter && f.matchn(str)) {
```
```suggestion
					if ((use_case_sensitive_filter && f.match(str)) || (!use_case_sensitive_filter && f.matchn(str))) {
```

AThousandShips commented:
You also need to register these enums in `_bind_methods`, see the other ones

AThousandShips commented:
You also need to document these methods, use `--doctool`, see [here](https://docs.godotengine.org/en/stable/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine)

DevilboxGames commented:
> You also need to document these methods, use `--doctool`, see [here](https://docs.godotengine.org/en/stable/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine)

I've updated the doc xml now, is this correct? The `--doctool` tool didn't generate any additional entries in it for the enum or methods.

 Documentation isn't my forte so the wording might need tweaking.

YuriSizov commented on doc/classes/FileDialog.xml:
```diff
@@ -133,6 +136,15 @@
 		<constant name="ACCESS_FILESYSTEM" value="2" enum="Access">
 			The dialog allows accessing files on the whole file system.
 		</constant>
+		<constant name="CASE_SENSITIVITY_TRUE" value="0" enum="CaseSensitivity">
+			The dialog will use a case sensitive check against the filters when saving.
+		</constant>
+		<constant name="CASE_SENSITIVITY_FALSE" value="1" enum="CaseSensitivity">
+			The dialog will use not a case sensitive check against the filters when saving.
+		</constant>
```
Generally speaking these are pretty weird names for constants. It's better to be explicit, e.g. `CASE_SENSITIVITY_SENSITIVE`/`CASE_SENSITIVITY_INSENSITIVE`.

YuriSizov commented on scene/gui/file_dialog.cpp:
```diff
@@ -1044,6 +1048,7 @@ void FileDialog::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::STRING, "root_subfolder"), "set_root_subfolder", "get_root_subfolder");
 	ADD_PROPERTY(PropertyInfo(Variant::PACKED_STRING_ARRAY, "filters"), "set_filters", "get_filters");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "show_hidden_files"), "set_show_hidden_files", "is_showing_hidden_files");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "case_sensitive_filter", PROPERTY_HINT_ENUM, "True,False,OS Default"), "set_case_sensitive_filter", "get_case_sensitive_filter");
```
Same for these strings.

AThousandShips commented on doc/classes/FileDialog.xml:
```diff
@@ -133,6 +136,15 @@
 			The dialog allows accessing files on the whole file system.
 		</constant>
+		<constant name="CASE_SENSITIVITY_SENSITIVE" value="0" enum="CaseSensitivity">
+			The dialog will use a case sensitive check against the filters when saving.	
```
```suggestion
			The dialog will use a case sensitive check against the filters when saving.
```
Stray whitespace

AThousandShips commented on scene/gui/file_dialog.cpp:
```diff
@@ -1043,6 +1047,7 @@ void FileDialog::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "access", PROPERTY_HINT_ENUM, "Resources,User Data,File System"), "set_access", "get_access");
 	ADD_PROPERTY(PropertyInfo(Variant::STRING, "root_subfolder"), "set_root_subfolder", "get_root_subfolder");
 	ADD_PROPERTY(PropertyInfo(Variant::PACKED_STRING_ARRAY, "filters"), "set_filters", "get_filters");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "case_sensitive_filter", PROPERTY_HINT_ENUM, "Sensitive,Insensitive,OS Default"), "set_case_sensitive_filter", "get_case_sensitive_filter");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::INT, "case_sensitive_filter", PROPERTY_HINT_ENUM, "Case Sensitive,Case Insensitive,OS Default"), "set_case_sensitive_filter", "get_case_sensitive_filter");
```
I'd say this is clearer

AThousandShips replied:
Or, by some standards:
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::INT, "case_sensitive_filter", PROPERTY_HINT_ENUM, "Case-Sensitive,Case-Insensitive,OS Default"), "set_case_sensitive_filter", "get_case_sensitive_filter");
```

AThousandShips commented on doc/classes/FileDialog.xml:
```diff
@@ -57,6 +57,9 @@
 			[b]Warning:[/b] Currently, in sandboxed environments such as Web builds or sandboxed macOS apps, FileDialog cannot access the host file system. See [url=https://github.com/godotengine/godot-proposals/issues/1123]godot-proposals#1123[/url].
 		</member>
+		<member name="case_sensitive_filter" type="int" setter="set_case_sensitive_filter" getter="get_case_sensitive_filter" enum="FileDialog.CaseSensitivity" default="2">
+			Whether the file type filters are case sensitive. See [enum CaseSensitivity].
```
```suggestion
			Whether the file type filters are case-sensitive. See [enum CaseSensitivity].
```
Seems to be the standard spelling, same below

AThousandShips replied:
Though we don't use it for `DirAccess` so let's keep it in the same style

AThousandShips replied:
We can look into standardising it later 🙂

DevilboxGames commented:
I noticed a merge conflict, so I've rebased and resolved it.

DevilboxGames commented:
It looks like one of the tests failed due to a ubuntu keyserver issue. Could a maintainer request a rerun of the failed test?

I appologise if the ping is innapropriate, but Yuri also requested requested @bruvzg to re-review the PR in December. Is there anything else which needs to be done before it can be approved for merging?

Calinou commented:
Tested locally on Linux (on a case-sensitive filesystem), I'm a bit confused about the behavior. The only behavior that seems desirable to me here is the one where case sensitive is set to **False**:

https://github.com/godotengine/godot/assets/180032/78d912ea-1b03-4c26-bcb4-f364bfacbac9

Otherwise, you get a double extension appended.

Also, the file filter is always case-insensitive regardless of this option, but I guess that's expected.

KoBeWi commented:
IMO this makes sense. The change only affects extension checking and I don't see why would you care about extension casing and make a double extension, like currently happens on mismatch.

The same change should be done in EditorFileDialog.

DevilboxGames commented:
> IMO this makes sense. The change only affects extension checking and I don't see why would you care about extension casing and make a double extension, like currently happens on mismatch.
> 
> The same change should be done in EditorFileDialog.

Good shout, I've added the change to EditorFileDialog too, I also noticed `FileDialog::_native_dialog_cb` needed the change too so I've dropped it in there.

KoBeWi commented:
Needs to be squashed into 1 commit.

DevilboxGames commented:
Commits squashed 👍

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
